File: src/main/java/leetcode/YY20/MM04/DD11_DiameterOfBinaryTree.java
package leetcode.YY20.MM04;

import leetcode.TreeNode;

public class DD11_DiameterOfBinaryTree {
    public static void main(String[] args) {
        TreeNode node = new TreeNode(1);
        node.left = new TreeNode(2);
        node.right = new TreeNode(3);
        node.left.left = new TreeNode(4);
        node.left.right = new TreeNode(5);

        System.out.println(diameterOfBinaryTree(node));
    }

    public static int diameterOfBinaryTree(TreeNode root) {
        if (root == null) return 0;

        int outer = getMaxLength(root.left) + getMaxLength(root.right);
        int inner = Math.max(diameterOfBinaryTree(root.left), diameterOfBinaryTree(root.right));
        return Math.max(outer, inner);
    }

    private static int getMaxLength(TreeNode node) {
        return (node == null) ? 0 : 1 + Math.max(getMaxLength(node.left), getMaxLength(node.right));
    }
}

File: src/main/kotlin/leetcode/YY20/MM04/DD11_DiameterOfBinaryTree.kt
import leetcode.TreeNode
import kotlin.math.max

fun main() {
    val node = TreeNode(1)
    node.left = TreeNode(2)
    node.right = TreeNode(3)
    node.left!!.left = TreeNode(4)
    node.left!!.right = TreeNode(5)

    println(diameterOfBinaryTree(node))
}

private fun diameterOfBinaryTree(root: TreeNode?): Int {
    if (root == null) return 0
    val outer = getMaxLength(root.left) + getMaxLength(root.right)
    val inner = max(diameterOfBinaryTree(root.left), diameterOfBinaryTree(root.right))
    return max(outer, inner)
}

private fun getMaxLength(node: TreeNode?): Int {
    return if (node == null) 0 else 1 + max(getMaxLength(node.left), getMaxLength(node.right))
}

/*internal class leetcode.TreeNode internal constructor(var value: Int) {
    var left: leetcode.TreeNode? = null
    var right: leetcode.TreeNode? = null
}*/
